File: processing/geodienste_api.py
import os
import json
import time
from datetime import datetime
import logging
import httpx
from processing.constants import CANTONS, TOPICS

GEODIENSTE_EXPORT_ERROR_INVALID_TOKEN = "Data export information not found. Invalid token?"
GEODIENSTE_EXPORT_ERROR_PENDING = (
    "Cannot start data export because there is another data export pending"
)
GEODIENSTE_EXPORT_ERROR_UNEXPECTED = (
    "An unexpected error occurred. Please try again by starting a new data export."
)
GEODIENSTE_EXPORT_STATUS_QUEUED = "queued"
GEODIENSTE_EXPORT_STATUS_WORKING = "working"
GEODIENSTE_EXPORT_STATUS_FAILED = "failed"
GEODIENSTE_EXPORT_STATUS_SUCCESS = "success"


class GeodiensteApi:
    """Handles all calls to the geodienste API"""

    GEODIENSTE_BASE_URL = "https://geodienste.ch/"
    GEODIENSTE_INFO_URL = (
        GEODIENSTE_BASE_URL
        + "/info/services.json?"
        + "base_topics={base_topics}&topics={topics}&cantons={cantons}&language=de"
    )
    GEODIENSTE_DOWNLOAD_URL = GEODIENSTE_BASE_URL + "/downloads/{topic}/{token}/export.json"
    GEODIENSTE_STATUS_URL = GEODIENSTE_BASE_URL + "/downloads/{topic}/{token}/status.json"

    SLEEP_TIME = 60

    def __init__(self):
        pass

    def _get_client(self):
        auth = httpx.BasicAuth(username=os.environ["AuthUser"], password=os.environ["AuthPw"])
        return httpx.Client(auth=auth)

    def request_topic_info(self, client):
        """Request the topic information from geodienste.ch"""
        if client is None:
            client = self._get_client()
        cantons = ",".join(CANTONS)
        base_topics = ",".join([topic["base_topic"] for topic in TOPICS])
        topics = ",".join([topic["topic"] for topic in TOPICS])
        url = self.GEODIENSTE_INFO_URL.format(
            base_topics=base_topics, topics=topics, cantons=cantons
        )
        response = client.get(url)
        if response.status_code != httpx.codes.OK:
            logging.error(
                "Fehler beim Abrufen der Themeninformationen von geodienste.ch: %s  - %s",
                response.status_code,
                response.reason_phrase,
            )
            return []

        return json.loads(response.text)["services"]

    def start_export(self, topic, canton, token, start_time=datetime.now(), client=None):
        """Starts the data export of a specific topic and token"""
        logging.info("Starten des Datenexports für %s (%s)...", topic, canton)
        if client is None:
            client = self._get_client()
        url = self.GEODIENSTE_DOWNLOAD_URL.format(topic=topic, token=token)
        response = client.get(url)
        if response.status_code == httpx.codes.NOT_FOUND:
            message = json.loads(response.text)
            if message.get("error") == GEODIENSTE_EXPORT_ERROR_PENDING:
                start_time_diff = datetime.now() - start_time
                if start_time_diff.seconds > 60 * 10:
                    logging.error("Es läuft bereits ein anderer Export. Zeitlimite überschritten.")
                    return response

                logging.info("Es läuft gerade ein anderer Export. Versuche es in 1 Minute erneut.")
                time.sleep(self.SLEEP_TIME)
                return self.start_export(topic, canton, token, start_time, client)
        return response

    def check_export_status(self, topic, canton, token, start_time=datetime.now(), client=None):
        """Checks the export status of a specific topic and token"""
        logging.info("Überprüfen des Exportstatus für %s (%s)...", topic, canton)
        if client is None:
            client = self._get_client()
        url = self.GEODIENSTE_STATUS_URL.format(topic=topic, token=token)
        response = client.get(url)
        if response.status_code == httpx.codes.OK:
            message = json.loads(response.text)
            if (
                message.get("status") == GEODIENSTE_EXPORT_STATUS_QUEUED
                or message.get("status") == GEODIENSTE_EXPORT_STATUS_WORKING
            ):
                status = (
                    "in Bearbeitung"
                    if message.get("status") == GEODIENSTE_EXPORT_STATUS_WORKING
                    else "in Warteschlange"
                )
                start_time_diff = datetime.now() - start_time
                if start_time_diff.seconds > 60 * 10:
                    logging.error("Zeitlimite überschritten. Status ist %s", status)
                    return response

                logging.info("Export ist %s. Versuche es in 1 Minute erneut.", status)
                time.sleep(self.SLEEP_TIME)
                return self.check_export_status(topic, canton, token, start_time, client)
        return response

File: Geodatenbezug/GeodiensteApi.cs
﻿using System.IO.Compression;
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using Geodatenbezug.Models;
using Microsoft.Extensions.Logging;
using Polly;

namespace Geodatenbezug;

/// <summary>
/// Accesses the geodienste.ch API.
/// </summary>
#pragma warning disable SA1009 // Closing parenthesis should be spaced correctly
public class GeodiensteApi(ILogger<GeodiensteApi> logger, IHttpClientFactory httpClientFactory) : IGeodiensteApi
#pragma warning restore SA1009 // Closing parenthesis should be spaced correctly
{
    private const string GeodiensteBaseUrl = "https://geodienste.ch";

    /// <summary>
    /// Timeouts the execution for 1 minute before retrying.
    /// </summary>
    public virtual TimeSpan GetWaitDuration() => TimeSpan.FromMinutes(1);

    /// <inheritdoc />
    public async Task<List<Topic>> RequestTopicInfoAsync()
    {
        try
        {
            using var httpClient = httpClientFactory.CreateClient(nameof(GeodiensteApi));
            var cantons = string.Join(",", Enum.GetValues(typeof(Canton)).Cast<Canton>().Select(e => e.ToString()));
            var baseTopics = string.Join(",", Enum.GetValues(typeof(BaseTopic)).Cast<BaseTopic>().Select(e => e.ToString()));
            var topics = string.Join(",", Enum.GetValues(typeof(BaseTopic)).Cast<BaseTopic>().Select(e => e.ToString() + "_v2_0"));
            var url = $"{GeodiensteBaseUrl}/info/services.json?base_topics={baseTopics}&topics={topics}&cantons={cantons}&language=de";
            logger.LogInformation($"Rufe die Themeninformationen ab: {url}");

            var infoData = await httpClient.GetFromJsonAsync<GeodiensteInfoData>(url).ConfigureAwait(false);
            return infoData.Services;
        }
        catch (HttpRequestException ex)
        {
            logger.LogError(ex, $"Fehler beim Abrufen der Themeninformationen von geodienste.ch: {ex.Message}");
            return [];
        }
    }

    /// <inheritdoc />
    public async Task<HttpResponseMessage> StartExportAsync(Topic topic)
    {
        var token = GetToken(topic.BaseTopic, topic.Canton);
        var url = $"{GeodiensteBaseUrl}/downloads/{topic.BaseTopic}/{token}/export.json";
        logger.LogInformation($"{topic.TopicTitle} ({topic.Canton}): Starte den Datenexport mit {url}");
        using var httpClient = httpClientFactory.CreateClient(nameof(GeodiensteApi));

        var retryPolicy = Policy
            .HandleResult<HttpResponseMessage>((response) =>
            {
                if (response.StatusCode == HttpStatusCode.NotFound)
                {
                    var jsonString = response.Content.ReadAsStringAsync().ConfigureAwait(false).GetAwaiter().GetResult();
                    var errorResponse = JsonSerializer.Deserialize<GeodiensteExportError>(jsonString);
                    return errorResponse.Error.Equals(GeodiensteExportError.Pending, StringComparison.Ordinal);
                }

                return false;
            })
            .WaitAndRetryAsync(10, retryAttempt => GetWaitDuration(), (result, timeSpan, retryCount, context) =>
                {
                    var response = result.Result;
                    var jsonString = response.Content.ReadAsStringAsync().Result;
                    var errorResponse = JsonSerializer.Deserialize<GeodiensteExportError>(jsonString);
                    if (retryCount < 10)
                    {
                        logger.LogInformation($"{topic.TopicTitle} ({topic.Canton}): Es läuft gerade ein anderer Export. Versuche es in 1 Minute erneut.");
                    }
                    else
                    {
                        logger.LogError($"{topic.TopicTitle} ({topic.Canton}): Es läuft bereits ein anderer Export. Zeitlimite überschritten.");
                    }
                });

        return await retryPolicy.ExecuteAsync(async () =>
        {
            return await httpClient.GetAsync(url).ConfigureAwait(false);
        }).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task<HttpResponseMessage> CheckExportStatusAsync(Topic topic)
    {
        var token = GetToken(topic.BaseTopic, topic.Canton);
        var url = $"{GeodiensteBaseUrl}/downloads/{topic.BaseTopic}/{token}/status.json";
        logger.LogInformation($"{topic.TopicTitle} ({topic.Canton}): Prüfe den Status des Datenexports mit {url}");
        using var httpClient = httpClientFactory.CreateClient(nameof(GeodiensteApi));

        var retryPolicy = Policy
            .HandleResult<HttpResponseMessage>((response) =>
            {
                if (response.StatusCode == HttpStatusCode.OK)
                {
                    var jsonString = response.Content.ReadAsStringAsync().ConfigureAwait(false).GetAwaiter().GetResult();
                    var statusResponse = JsonSerializer.Deserialize<GeodiensteStatusSuccess>(jsonString);
                    return statusResponse.Status == GeodiensteStatus.Queued || statusResponse.Status == GeodiensteStatus.Working;
                }

                return false;
            })
            .WaitAndRetryAsync(10, retryAttempt => GetWaitDuration(), (result, timeSpan, retryCount, context) =>
            {
                var response = result.Result;
                var jsonString = response.Content.ReadAsStringAsync().Result;
                var statusResponse = JsonSerializer.Deserialize<GeodiensteStatusSuccess>(jsonString);
                var statusString = statusResponse.Status == GeodiensteStatus.Queued ? "in der Warteschlange" : "in Bearbeitung";
                if (retryCount < 10)
                {
                    logger.LogInformation($"{topic.TopicTitle} ({topic.Canton}): Export ist {statusString}. Versuche es in 1 Minute erneut.");
                }
                else
                {
                    logger.LogError($"{topic.TopicTitle} ({topic.Canton}): Zeitlimite überschritten. Status ist {statusString}");
                }
            });

        return await retryPolicy.ExecuteAsync(async () =>
        {
            return await httpClient.GetAsync(url).ConfigureAwait(false);
        }).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task<string> DownloadExportAsync(string downloadUrl, string destinationPath)
    {
        logger.LogInformation($"Lade die Daten herunter {downloadUrl}");
        Directory.CreateDirectory(destinationPath);
        var downloadedFilePath = string.Empty;
        using var httpClient = httpClientFactory.CreateClient(nameof(GeodiensteApi));
        using var stream = await httpClient.GetStreamAsync(downloadUrl).ConfigureAwait(false);
        using var archive = new ZipArchive(stream);
        foreach (var entry in archive.Entries)
        {
            if (entry.Name.EndsWith(".gpkg", StringComparison.OrdinalIgnoreCase))
            {
                downloadedFilePath = Path.Combine(destinationPath, entry.Name);
                entry.ExtractToFile(downloadedFilePath, overwrite: true);
                break;
            }
        }

        if (string.IsNullOrEmpty(downloadedFilePath))
        {
            throw new FileNotFoundException("Keine GeoPackage-Datei im Archiv gefunden.");
        }

        return downloadedFilePath;
    }

    /// <summary>
    /// Gets the token for the given topic and canton from the environment variables.
    /// </summary>
    protected internal virtual string GetToken(BaseTopic baseTopic, Canton canton)
    {
        var topicTokens = Environment.GetEnvironmentVariable("tokens_" + baseTopic.ToString());
        if (string.IsNullOrEmpty(topicTokens))
        {
            throw new InvalidOperationException($"No tokens available for topic {baseTopic}");
        }

        return topicTokens.ExtractValueByKey(canton.ToString());
    }
}
